libraries/vault_auth.rb

```diff
@@ -0,0 +1,140 @@
+#
+# Cookbook: hashicorp-vault
+# License: Apache 2.0
+#
+# Copyright 2016, Parallels IP Holdings GmbH.
+#
+require 'poise'
+
+# Inspiration from sethvargo's gist
+module VaultCookbook
+  module Resource
+    # A `vault_auth` resource for reading secrets out of Vault
+    # @action auth
+    # @provides vault_auth
+    # @since 2.5
+    class VaultAuth < Chef::Resource
+      include Poise(fused: true)
+      provides(:vault_auth)
+      actions(:auth)
+      default_action(:auth)
+
+      # @!attribute address
+      # The address of Vault server, for example: "http://127.0.0.1:8200"
+      # supported.
+      # @return [String]
+      attribute(:address, kind_of: String, name_attribute: true)
+
+      # @!attribute token
+      # The Vault token. If specified, the resource will try to renew this
+      # token first. If the renewal fails, the regular auth will be performed.
+      # @return [String]
+      attribute(:token, kind_of: [String, NilClass], default: nil)
+
+      # @!attribute type
+      # The authentication type (auth backend). Currently, only "approle" is
+      # supported.
+      # @return [String]
+      attribute(:type, equal_to: ['approle'], default: 'approle')
```

Comment:
It doesn't look like this attribute is being used anywhere, but I can see its value when multiple types of auth backends are supported (e.g. using it in a `case` statement). I think it would be great if this provider could be flexible enough to run a script to fetch a token. If it could, it would be one way for me to leverage this provider with my [SmartProxy plugin](https://github.com/theforeman/smart_proxy_vault) which is what handles Vault auth for our Chef nodes.